many tests passing - rails4

unorthodoxgeek opened the pull request.

dLobatog committed:
Update gems to a Rails 4 compatible version

dLobatog committed:
Remove conditions,order,limit from has_many relations

unorthodoxgeek committed:
refs #3157 - adding protected_attributes gem

Conflicts:
	app/models/auth_source.rb
	app/models/domain.rb
	app/models/image.rb
	app/models/lookup_value.rb
	config/routes/test.rb

dLobatog committed:
Fixes #11924 - Substitute .scoped by .where(nil) to force return relation

On Rails 4 .scoped is deprecated. Calling .all on the model returns the
equivalent ActiveRecord relation object on Rails 4, but on Rails 3 it
returns an Array right away.

A proper replacement we can use is where(nil) - it's ugly but it returns
the same relation in both Rails 3 and 4. There are a couple of fixes too
on models such as bookmark.rb which return where conditions ({} and
similar) instead of a relation, which is also a deprecated behavior.

We could possibly substitute these by .all after the Rails 4 migration
if they feel too 'unidiomatic'.

dLobatog committed:
Refs #10409 - Rails 4 compatible changes with Rails 3

This includes some changes that can be applied to Rails 3 without
any issue and don't follow any particular pattern/deprecation, In
general these are minor things we have to change. Other PRs with
retrocompatible changes that aim to fix particular deprecations
have been submitted separately.

Sorry for the not-very-descriptive commit!

dLobatog committed:
Use explicit friendly ID search

dLobatog committed:
Fixes #11986 - Specify PUT method on form_for tags

Rails 4 will default to PATCH which we have not implemented in our
routes. This will cause at least one failure on taxonomies integration
tests. It's the only one our test suite catches, however I'd expect
other forms to fail on this when tested manually. We can keep this
issue number as a reference and add commits with integration tests
& fixes where needed.

dLobatog committed:
Protected attributes automatically added through AccessibleAttributes

AccessibleAttributes pulls in all possible associations and models and
sets them as attr_accessible for the model. Where needed, additional
attr_accessible attributes have been added.

dLobatog committed:
Fixes #12021 - Use .to_param to find parent object in LookupKeys controller

On Rails 4, .find will not default to .friendly.find so find_hostgroup,
find_environment and find_host will fail for non numeric IDs. However, we can
use from_param to find these objects. This is something we can do both on
Rails 3 and Rails 4, whereas using .friendly isn't an option until friendly_id
 5.0, which depends on Rails 4.

dLobatog committed:
Refs #11813 - Use new engines listing syntax in Apipie initializer

dLobatog committed:
Counter cache columns fix

dLobatog committed:
Rails 4 non backwards compatible syntax changes

Changes that have to deal with how some of the internal Rails objects
need a new syntax, like generating routes, exceptions, form builders.
None of these are Rails 3 compatible.

dLobatog committed:
url_for doesn't append port when protocol is specified

Some of our tests are checking if port 80 or 443 is included in the URL
generated by lib/foreman/renderer.rb. However url_for has changed in
Rails 4 and now it does not append the port even if explicitly added.

If the protocol is specified, as it is in this case, then the url
generated will just have the protocol (http or https) but not the port
(80 or 443)

unorthodoxgeek committed:
many tests passing

dLobatog commented on app/models/config_group_class.rb:
```diff
@@ -10,6 +10,6 @@ class ConfigGroupClass < ActiveRecord::Base
   validates :puppetclass_id, :presence => true
-  validates :config_group_id, :presence => true,
-                              :uniqueness => {:scope => :puppetclass_id}
+  #validates :config_group_id, :presence => true,
```
These two fields can be changed to `validate :puppetclass` and `validate :config_group` and tests will pass.

dLobatog commented on app/models/user_mail_notification.rb:
```diff
@@ -3,7 +3,7 @@ class UserMailNotification < ActiveRecord::Base
   include AccessibleAttributes
 
-  validates :user_id, :presence => true
+  #validates :user_id, :presence => true
```
Same here, no need to remove the validation, just remove the "_id"

dLobatog commented on app/models/lookup_keys/lookup_key.rb:
```diff
@@ -57,7 +57,7 @@ def self.inherited(child)
   alias_attribute :override_values, :lookup_values
   alias_attribute :override_value_ids, :lookup_value_ids
-  attr_accessible :lookup_values_attributes, :is_param, :id, :variable, :key
+  attr_accessible :lookup_values_attributes, :is_param, :id, :variable, :key, :puppetclass_id
```
I've put this one on variable_lookup_key , otherwise puppetclass_id might not even be an attribute

unorthodoxgeek replied:
ok

dLobatog committed:
Validate object instead of _id column in join tables

On Rails 4 such validations will fail when we try to create objects
without explicitely assigning the id.
Puppetclass.new(:config_group => config_group) would fail even for a
valid config group, :config_group_id => config_group.id should be used
instead.

To avoid that, we validate the object, not the ID column

dLobatog committed:
Fixes #12098 - Dangling Fog.mock makes compute_resource_vms test fail

test/unit/compute_resource_test.rb contains a Fog.mock! without a
Fog.unmock!. This causes
test/functional/compute_resource_vms_controller_test.rb to fail only
when ran through rake test on Rails 4. Fix is as simple as adding the
missing unmock!.

dLobatog committed:
Fixes #12099 - Template locked? relying on rake makes tests fail

app/models/template.rb has a method locked? that relies on
!Foreman.in_rake? to bypass the locked validation when running some
operations through Rake such as migrations or cloning templates. On
Rails 4 rake test is also interpreted as being 'in rake' and tests fail
when running them through rake test but not when running template tests
individually.

dLobatog committed:
RESET AND MAKE MEANINGFUL COMMITS
